src/main/kotlin/com/autonomousapps/internal/parse/KotlinBuildScriptDependenciesRewriter.kt

```diff
@@ -0,0 +1,114 @@
+package com.autonomousapps.internal.parse
+
+import cash.grammar.kotlindsl.parse.KotlinParseException
+import cash.grammar.kotlindsl.parse.Parser
+import cash.grammar.kotlindsl.utils.Blocks.isDependencies
+import cash.grammar.kotlindsl.utils.CollectingErrorListener
+import cash.grammar.kotlindsl.utils.DependencyExtractor
+import cash.grammar.kotlindsl.utils.Whitespace
+import com.autonomousapps.internal.advice.AdvicePrinter
+import com.autonomousapps.internal.utils.ifNotEmpty
+import com.autonomousapps.model.Advice
+import com.squareup.cash.grammar.KotlinParser
+import com.squareup.cash.grammar.KotlinParserBaseListener
+import org.antlr.v4.runtime.CharStream
+import org.antlr.v4.runtime.CommonTokenStream
+import org.antlr.v4.runtime.TokenStreamRewriter
+import java.nio.file.Path
+
+/**
+ * Rewrites a Kotlin build script with advice.
+ *
+ * Logic:
+ * - Ignores buildscript dependencies
+ * - When enter each declaration, removes or updated it if based on the advice. See `handleDeclaration` function
+ * - At the end of the script, add any new dependencies
+ */
+internal class KotlinBuildScriptDependenciesRewriter(
+  private val input: CharStream,
+  private val tokens: CommonTokenStream,
+  private val errorListener: CollectingErrorListener,
+  private val advice: Set<Advice>,
+  private val printer: AdvicePrinter,
+  /** Reverse map from custom representation to standard. */
+  private val reversedDependencyMap: (String) -> String,
+) : BuildScriptDependenciesRewriter, KotlinParserBaseListener() {
+  private val rewriter = TokenStreamRewriter(tokens)
+  private val indent = Whitespace.computeIndent(tokens, input)
+
+  private val dependencyExtractor = DependencyExtractor(
+    input = input,
+    tokens = tokens,
+    indent = indent,
+  )
+
+  /**
+   * Returns build script with advice applied.
+   *
+   * Throws [KotlinParseException] if the script has some idiosyncrasy that impairs parsing.
+   *
+   */
+  @Throws(KotlinParseException::class)
+  override fun rewritten(): String {
+    errorListener.getErrorMessages().ifNotEmpty {
+      throw KotlinParseException.withErrors(it)
+    }
+
+    return rewriter.text
+  }
+
+  override fun enterNamedBlock(ctx: KotlinParser.NamedBlockContext) {
+    dependencyExtractor.onEnterBlock()
+
+    if (ctx.isDependencies) {
+      handleDependencies(ctx)
+    }
+  }
+
+  override fun exitNamedBlock(ctx: KotlinParser.NamedBlockContext) {
+    dependencyExtractor.onExitBlock()
+
+    // TODO(swipawiwat): all new dependencies to the end of the block if not in buildscript block
+  }
+
+  override fun exitScript(ctx: KotlinParser.ScriptContext?) {
```

Comment:
minor nits: (1) we can add imports for all contexts so we need need to qualify them all with `KotlinParser.` and (2) they're all non-null, but written in Java, so we have to manually remove the `?`.

Reply:
👍 I think it comes from auto complete. I will fix them :)

Reply:
Fixed!